File: cplusplus/src/cert.c
#include "cert.h"

#include <string.h>

#include "constants.h"
#include "external.h"
#include "state.h"

size_t calculate_subject_size(const size_t data_size) {
  size_t max_size = UINT32_MAX - PUBLIC_KEY_SIZE;
  size_t size = data_size > max_size ? max_size : data_size;
  return size + PUBLIC_KEY_SIZE;
}

void calculate_subject(uint8_t *subject, const size_t subject_size,
                       const uint8_t *public_key, const uint8_t *data) {
  size_t key_offset = subject_size - PUBLIC_KEY_SIZE;
  memmove(subject, data, key_offset);
  memmove(subject + key_offset, public_key, PUBLIC_KEY_SIZE);
}

bool sign_subject(uint8_t *signature, uint8_t *state, const uint8_t *subject,
                  const size_t subject_size) {
  return sign(signature, get_identity_key_pair(state), subject, subject_size);
}

bool certify_data_ownership(uint8_t *signature, uint8_t *state,
                            const uint8_t *owner_public_key,
                            const uint8_t *data, const size_t data_size) {
  size_t subject_size = calculate_subject_size(data_size);
  uint8_t subject[subject_size];
  calculate_subject(subject, subject_size, owner_public_key, data);
  return sign_subject(signature, state, subject, subject_size);
}

bool certify_identity_ownership(uint8_t *signature, uint8_t *state,
                                const uint8_t *owner_public_key) {
  return sign_subject(signature, state, owner_public_key, PUBLIC_KEY_SIZE);
}

bool verify_data_ownership(const uint8_t *owner_public_key, const uint8_t *data,
                           const size_t data_size,
                           const uint8_t *certifier_public_key,
                           const uint8_t *signature) {
  size_t subject_size = calculate_subject_size(data_size);
  uint8_t subject[subject_size];
  calculate_subject(subject, subject_size, owner_public_key, data);
  return verify(certifier_public_key, signature, subject, subject_size);
}

bool verify_identity_ownership(const uint8_t *owner_public_key,
                               const uint8_t *certifier_public_key,
                               const uint8_t *signature) {
  return verify(certifier_public_key, signature, owner_public_key,
                PUBLIC_KEY_SIZE);
}

File: rust/src/cert.rs
use alloc::{vec, vec::Vec};

use crate::{
    constants::PUBLIC_KEY_SIZE,
    external::{sign, verify},
    state::get_identity_key_pair,
    types::{PublicKey, Signature, State},
};

fn create_subject(data: &[u8]) -> Vec<u8> {
    let max_size = (u32::MAX as usize) - PUBLIC_KEY_SIZE;
    let data_size = if data.len() > max_size {
        max_size
    } else {
        data.len()
    };
    vec![0; data_size + PUBLIC_KEY_SIZE]
}

fn calculate_subject(public_key: &PublicKey, data: &[u8]) -> Vec<u8> {
    let mut subject = create_subject(data);
    let key_offset = subject.len() - PUBLIC_KEY_SIZE;
    subject[..key_offset].copy_from_slice(&data[..key_offset]);
    subject[key_offset..].copy_from_slice(public_key);
    subject
}

fn sign_subject(signature: &mut Signature, state: &State, subject: &[u8]) -> bool {
    sign(signature, get_identity_key_pair(state), subject)
}

pub fn certify_data_ownership(
    signature: &mut Signature,
    state: &State,
    owner_public_key: &PublicKey,
    data: &[u8],
) -> bool {
    let subject = calculate_subject(owner_public_key, data);
    sign_subject(signature, state, &subject)
}

pub fn certify_identity_ownership(
    signature: &mut Signature,
    state: &State,
    owner_public_key: &PublicKey,
) -> bool {
    sign_subject(signature, state, owner_public_key)
}

pub fn verify_data_ownership(
    owner_public_key: &PublicKey,
    data: &[u8],
    certifier_public_key: &PublicKey,
    signature: &Signature,
) -> bool {
    let subject = calculate_subject(owner_public_key, data);
    verify(certifier_public_key, signature, &subject)
}

pub fn verify_identity_ownership(
    owner_public_key: &PublicKey,
    certifier_public_key: &PublicKey,
    signature: &Signature,
) -> bool {
    verify(certifier_public_key, signature, owner_public_key)
}
